### PR TITLE
Fix filtering properties in TypeDescriptor

### LINKED ISSUE #1181: WindowsBase provided with Net Core 3 for WPF applications contains a bug
**Description:**

There is a method in .Net Framework TypeDescriptor.GetProperties(Object, Attribute[])

This particular override allows you to pass PropertyFilterAttribute as an item of array. Then method filter your properties according to enum value provided with attribute.

It seems that in .Net Core 3 libs that was built to help us make WPF appliations with Net Core this behavior has changed. In net core this method checks is property actually has this attribute on it with provided enum values. This doesn't match previous behavior at all.

**Steps to reproduce:**

1. Create Solution
2. Create .Net Framework Console Application
3. Use following code
```
class Program
{
    public class POCO
    {
        public string StringPrpp { get; set; }
        public int IntProp { get; set; }
        public double DoubleProp { get; set; }
    }

    private static readonly Attribute[] PropertyFilter = new Attribute[] { new PropertyFilterAttribute(PropertyFilterOptions.SetValues | PropertyFilterOptions.UnsetValues | PropertyFilterOptions.Valid) };

    static void Main(string[] args)
    {
        Console.WriteLine(TypeDescriptor.GetProperties(typeof(POCO), PropertyFilter).Count);
        Console.ReadLine();
    }
}
```

4. If you run this code you will see "3" on the screen, meaning that TypeDescriptor found 3 properties
5. Create Net Core **WPF!** application _(Looks like PropertyFilter declared somewhere in WindowsBase, and this bug happens when we use WindowsBase provided with Net Core which I cant reference without creating net core WPF application)_
Change App.cs with following code
```
 public partial class App : Application
    {
        public class POCO
        {
            public string StringPrpp { get; set; }
            public int IntProp { get; set; }            
            public double DoubleProp { get; set; }
        }

        private static readonly Attribute[] PropertyFilter = new Attribute[] { new PropertyFilterAttribute(PropertyFilterOptions.SetValues | PropertyFilterOptions.UnsetValues | PropertyFilterOptions.Valid) };
               
        public App()
        {
            var count = TypeDescriptor.GetProperties(typeof(POCO), PropertyFilter).Count;
        }

    }
```

Do not forget to add `using System.ComponentModel;`
6. Put a breakpoint after count. Or write value to output. You will see that count equals "0" (zero);
In net core application you also can add [PropertyFilter(PropertyFilterOptions.SetValues | PropertyFilterOptions.UnsetValues | PropertyFilterOptions.Valid)] to any property and as a result count will be increased

### PR DESCRIPTION
**Pull request by svick**
Fixes https://github.com/microsoft/dotnet/issues/1181.

The explanation for this change (copied from https://github.com/microsoft/dotnet/issues/1181#issuecomment-593974499):

> I believe the problem is in `TypeDescriptor.ShouldHideMember`. [On .Net Framework, it calls `attribute.Match(memberAttribute)`](https://github.com/microsoft/referencesource/blob/a7bd324/System/compmod/system/componentmodel/TypeDescriptor.cs#L3312), while [on .Net Core, it calls `attribute.Equals(memberAttribute)` instead](https://github.com/dotnet/runtime/blob/19c11e6/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L2562) (presumably because [`Attribute.Match`](https://docs.microsoft.com/en-us/dotnet/api/system.attribute.match) didn't exist in .Net Core 1.0).
> 
> This would explain why the properties are returned on .Net Framework (because the attributes match), but not on .Net Core (because the attributes aren't equal).

### CODE REVIEW

File: src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs [hunk 1]
```diff
@@ -1141,6 +1141,30 @@ public void DerivedPropertyAttribute()
...
+            public override bool Equals(object value) => ReferenceEquals(this, value);
+
+            public override int GetHashCode() => throw new NotImplementedException();
+
+            public override bool Match(object value) => true;
+
+            public static readonly PropertyFilterAttribute Default = new PropertyFilterAttribute();
+        }
+
+        public class POCO
```
Comment (safern) [inline src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs:1161]: Can this class name be more descriptive? Maybe PropertyFilterAttributeMatchPoco?

File: src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs [hunk 2]
```diff
@@ -1141,6 +1141,30 @@ public void DerivedPropertyAttribute()
...
+            Assert.Equal(3, TypeDescriptor.GetProperties(typeof(POCO), new[] { new PropertyFilterAttribute() }).Count);
+        }
+
+        public sealed class PropertyFilterAttribute : Attribute
+        {
+            public override bool Equals(object value) => ReferenceEquals(this, value);
+
+            public override int GetHashCode() => throw new NotImplementedException();
+
+            public override bool Match(object value) => true;
```
Comment (safern) [inline src/libraries/System.ComponentModel.TypeConverter/tests/TypeDescriptorTests.cs:1156]: Should we add a test where `Match` returns false for a specific property to make sure we also filter negatively?

**Comment by ericstj** [comment]
@svick are you still interested in this fix, could you address @safern's feedback?

**Review by ericstj — APPROVED** [review]

**Comment by safern** [comment]
cc: @joperezr @eerhardt for the linker tests failure.

**Comment by eerhardt** [comment]
```
Could not copy "/Users/runner/work/1/s/artifacts/obj/System.Runtime.InteropServices/ref/net6.0-Release/System.Runtime.InteropServices.pdb" to "/Users/runner/work/1/s/artifacts/bin/System.Runtime.InteropServices/ref/net6.0-Release/System.Runtime.InteropServices.pdb". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/System.Runtime.InteropServices/ref/net6.0-Release/System.Runtime.InteropServices.pdb' because it is being used by another process.
```

This is odd - why are we re-building the ref file at all at this point in the build/test run? cc @ViktorHofer

**Comment by ViktorHofer** [comment]
Unfortunately I can't tell as there's an out-of-process restore/build happening which doesn't create a binlog:

image

**Comment by ViktorHofer** [comment]
My guess is that because of the non-incremental restore which updates the intermediate files, the build isn't incremental as the inputs are newer than the outputs. I would look into why the restore wasn't incremental but without a binlog this is probably impossible.

**Comment by ViktorHofer** [comment]
Some more observations:
- Linker tests don't use static graph restore which is used by the rest of the repository. This could cause some of the incremental issues that we are seeing.
- The shelving out into an Exec is problematic as we don't have any logging for it. I would recommend converting this into a normal MSBuild task so that cached nodes are reused (perf) which might also eliminate the issue with incremental builds. IIRC this design was chosen as there's no way to unset global properties in msbuild and the environment was supposed to be entirely clean but IMHO I wouldn't be concerned about any global properties being set as we hardly use them anymore.

**Comment by safern** [comment]
> IIRC this design was chosen as there's no way to unset global properties in msbuild and the environment was supposed to be entirely clean but IMHO I wouldn't be concerned about any global properties being set as we hardly use them anymore.

Yep, that was exactly why and that could also cause some race conditions we've seen when enabled parallelism on the tests which we had to disable in the past. I'm also leaned towards making this an MSBuild task.

**Comment by ericstj** [comment]
Normally we avoid rebuilding projects through using MSBuild's build cache.  Do we have to make this step happen through another root build call?  Can't we just invoke the appropriate builds on test projects as part of the same build invocation?

**Comment by ViktorHofer** [comment]
> Normally we avoid rebuilding projects through using MSBuild's build cache. Do we have to make this step happen through another root build call? Can't we just invoke the appropriate builds on test projects as part of the same build invocation?

That was my suggestion (see above). The reasoning back then to shelve out was to not pass in global properties which IMO isn't a concern as we don't have many global properties set anymore.

**Pull request merged**